### PR TITLE
Implement the read path of paged local block store

### PR DESCRIPTION
**Pull request by beinan**
### What changes are proposed in this pull request?

Implement the read path of pagedd local block store

### Why are the changes needed?

Please clarify why the changes are needed. For instance,
Improve the stability of worker
Prevent read amplification

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
Added new properties to enable the paged local block store

### CODE REVIEW

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 1]
```diff
@@ -146,7 +148,7 @@ public class DefaultBlockWorker extends AbstractWorker implements BlockWorker {
     this(new BlockMasterClientPool(),
         new FileSystemMasterClient(MasterClientContext
             .newBuilder(ClientContext.create(ServerConfiguration.global())).build()),
-        new Sessions(), new TieredBlockStore(), ufsManager);
+        new Sessions(), LocalBlockStore.create(ufsManager), ufsManager);
```
Comment (rongrong) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java:151]: I'm assuming that later this will be configurable.

Comment (beinan) [reply core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java:147]: done, made the instance type of LocalBlockStore configurable

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockMetaStore.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  @Override
+  public Map<String, List<String>> getLostStorage() {
+    return Collections.EMPTY_MAP;
+  }
+
+  @Override
+  public int getNumberOfBlocks() {
+    return mBlockPageMap.size();
```
[the remark this answers is not in the document]
Comment (beinan) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockMetaStore.java]: Good catch,  added a read lock here, thx!

File: core/server/worker/src/main/java/alluxio/worker/block/LocalBlockStore.java [hunk 3]
```diff
@@ -33,6 +41,26 @@
  */
 public interface LocalBlockStore
     extends SessionCleanable, Closeable {
+  /**
+   * @return the instance of LocalBlockStore
+   * @param ufsManager
```
Comment (apc999) [inline core/server/worker/src/main/java/alluxio/worker/block/LocalBlockStore.java:46]:
```suggestion
   * @param ufsManager ...
   * @return the instance of LocalBlockStore
```

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockReader.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A paged implementation of BlockReader interface. The read operations will fall back to the
+ * under storage when the requested data is not in the local storage.
+ */
+@NotThreadSafe
+public class PagedBlockReader extends BlockReader {
+
+  public static final ByteBuffer EMPTY_BYTE_BUFFER = ByteBuffer.allocate(0);
```
Comment (apc999) [inline core/server/worker/src/main/java/alluxio/worker/page/PagedBlockReader.java:45]: does this need to be public?

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockMetaStore.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  public static final BlockStoreLocation DEFAULT_BLOCK_STORE_LOCATION =
+      new BlockStoreLocation(DEFAULT_TIER, 0);
+  public static final String DEFAULT_DIR = "DefaultDir";
+  public static final DefaultStorageTierAssoc DEFAULT_STORAGE_TIER_ASSOC =
+      new DefaultStorageTierAssoc(ImmutableList.of("SSD"));
+
+  private final long mCapacity;
+  private final ReentrantReadWriteLock mBlockPageMapLock = new ReentrantReadWriteLock();
+  @GuardedBy("mBlockPageMapLock")
+  private Map<Long, Set<Long>> mBlockPageMap = new HashMap<>();
```
Comment (apc999) [inline core/server/worker/src/main/java/alluxio/worker/page/PagedBlockMetaStore.java:55]: final?

File: core/server/worker/src/main/java/alluxio/worker/block/LocalBlockStore.java [hunk 6]
```diff
@@ -183,6 +211,21 @@ BlockWriter getBlockWriter(long sessionId, long blockId)
...
+   * <p>
+   * The block reader will fetch the data from UFS when the data is not cached by the worker
+   *
+   * @param sessionId the id of the session to get the reader
+   * @param blockId the id of an existing block
+   * @param options the options for UFS fall-back
+   * @return a {@link BlockReader} instance on this block
+   */
+  default BlockReader createBlockReader(long sessionId, long blockId,
+                                        Protocol.OpenUfsBlockOptions options) {
```
Comment (apc999) [inline core/server/worker/src/main/java/alluxio/worker/block/LocalBlockStore.java:225]: shall we throw same exception as above method?

**Review by apc999 — APPROVED** [review]
I didn't go through impl details. but the structure / hierarchy look good to me.
Thanks!

File: core/server/worker/src/main/java/alluxio/worker/block/LocalBlockStore.java (hunk 3 above)
Comment (beinan) [reply core/server/worker/src/main/java/alluxio/worker/block/LocalBlockStore.java:46]: done

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockReader.java (hunk 4 above)
Comment (beinan) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockReader.java]: good catch, changed it private

File: core/server/worker/src/main/java/alluxio/worker/page/PagedBlockMetaStore.java (hunk 5 above)
Comment (beinan) [reply core/server/worker/src/main/java/alluxio/worker/page/PagedBlockMetaStore.java]: done

**Comment by beinan** [comment]
alluxio-bot, merge this please.

**Pull request merged**